nengo_loihi/splitter.py

```diff
@@ -1,543 +1,141 @@
+            self._seen_objects.add(ens)
+
+        # Step 3. Move learning ensembles (post and error) to host
+        for conn in network.all_connections:
+            pre = base_obj(conn.pre)
+            post = base_obj(conn.post)
+            if (conn.learning_rule_type is not None
+                    and isinstance(post, Ensemble)
+                    and post in self._chip_objects):
+                if network.config[post].on_chip:
+                    raise BuildError("Post ensemble (%r) of learned "
+                                     "connection (%r) must not be configured "
+                                     "as on_chip." % (post, conn))
+                self._chip_objects.remove(post)
+            elif (isinstance(post, LearningRule)
+                  and isinstance(pre, Ensemble)
+                  and pre in self._chip_objects):
+                if network.config[pre].on_chip:
+                    raise BuildError("Pre ensemble (%r) of error "
+                                     "connection (%r) must not be configured "
+                                     "as on_chip." % (pre, conn))
+                self._chip_objects.remove(pre)
+
+        # Step 4. Mark passthrough nodes for removal
+        if remove_passthrough:
+            passthroughs = set(
+                obj for obj in network.all_nodes if is_passthrough(obj))
+            ignore = self._seen_objects - self._chip_objects - passthroughs
+            self.passthrough_directive = convert_passthroughs(network, ignore)
         else:
-            self.moves[obj] = target
-
-    def remove(self, obj):
-        obj = base_obj(obj)
-        logger.debug("Removing %s", obj)
-        self.removes.append(obj)
-        if obj in self.adds:
-            del self.adds[obj]
-        elif obj in self.moves:
-            del self.moves[obj]
-
-
-def split(net, precompute, node_neurons, node_tau, remove_passthrough=False):
-    logger.info("Splitting model into host and chip parts")
-    networks = SplitNetworks(net, node_neurons=node_neurons,
-                             node_tau=node_tau)
-
-    # --- Step 1: place ensembles and nodes
-    place_nodes(networks)
-    place_ensembles(networks)
-
-    # --- Step 1b: remove passthrough nodes
-    if remove_passthrough:
-        conns = merge_passthrough_nodes(networks)
-    else:
-        conns = networks.original.all_connections
-
-    # --- Step 2: place simple connections
-    place_internetwork_connections(networks, conns)
-
-    # --- Step 3: split complex connections
-    split_host_to_chip_connections(networks, conns)
-    split_chip_to_host_connections(networks, conns)
-    split_host_to_learning_rules(networks, conns)
-
-    # --- Step 4: place precomputable parts of host
-    if precompute:
-        split_pre_from_host(networks)
-
-    # --- Step 5: place probes
-    place_probes(networks)
+            self.passthrough_directive = PassthroughDirective(
+                set(), set(), set())
```

Comment:
Seems a bit strange to import both `convert_passthroughs` and `PassthroughDirective`. I would pick one as the interface, and have it somehow able to return and empty directive in the case that `remove_passthrough == False`.

Reply:
Yeah it's also strange that I import `is_passthrough` and have to manually filter things before calling the function. But right now all this weirdness/friction is limited to 6 lines of code, and I have a feeling all of it will change when we redo the passthrough removal. Would it be better to define an `EmptyPassthroughDirective` within passthrough.py and then import+use that?

Reply:
I don't think that would make much of a difference. That's a fair point that it's only 6 lines and might change, and there's not an obvious way to do it better.